c10/cuda/CUDAFunctions.cpp

```diff
@@ -182,4 +182,40 @@ bool hasPrimaryContext(int64_t device_index) {
   return _internal::hasPrimaryContext(device_index);
 }
 
+std::atomic_int LastSavedDevice = -1;
```

Comment:
The device is thread local so this variable should be thread local as well. Otherwise multiple threads will use this single value to save their current thread.